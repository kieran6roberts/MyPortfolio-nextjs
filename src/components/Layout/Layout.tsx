import * as React from "react"
import Image from "next/image";
import Link from "next/link";
import { RiComputerLine } from "react-icons/ri";
import { AiOutlinePhone, AiFillGithub } from "react-icons/ai";
import { FaLinkedinIn } from "react-icons/fa";
import { motion as m, useViewportScroll, useTransform, useSpring } from "framer-motion";

export type LayoutProps = {
    children: JSX.Element | JSX.Element[]
}
    
export default function Layout({ children }: LayoutProps): React.ReactElement {
    const { scrollYProgress } = useViewportScroll();
    const [ isComplete, setIsComplete ] = React.useState(false);
    const yRange = useTransform(scrollYProgress, [0, 0.9], [0, 1]);
    const pathLength = useSpring(yRange, { stiffness: 400, damping: 90 });
    
    React.useEffect(() => {
        yRange.onChange(v => setIsComplete(v >= 1));
    }, [yRange, scrollYProgress]);

    return (
        <div className="flex flex-col justify-between font-mono">
            <section className="relative z-50">
            <svg className="block fixed bottom-8 md:bottom-16 right-0 md:right-12 h-12 w-12" 
            viewBox="0 0 60 60">
                <m.path
                fill="none"
                strokeWidth="4"
                stroke="black"
                strokeDasharray="0 1"
                d="M 0, 20 a 20, 20 0 1,0 40,0 a 20, 20 0 1,0 -40,0"
                style={{
                    pathLength,
                    rotate: 90,
                    translateX: 5,
                    translateY: 5,
                    scaleX: -1 // Reverse direction of line animation
                }}
                />
                <m.path
                fill="none"
                strokeWidth="5"
                stroke="white"
                d="M14,26 L 22,33 L 35,16"
                initial={false}
                strokeDasharray="0 1"
                animate={{ pathLength: isComplete ? 1 : 0 }}
                />
            </svg>
                <nav className="">
                    <div className="h-1 w-full bg-gradient-to-r from-purple-400 to-blue-800"/>
                        <div className="flex items-center py-8 px-6 md:px-16">
                            <div className="flex p-2 mr-4 cursor-pointer">
                                <Link href="/" passHref>
                                    <a className="flex align-center focus:outline-none focus:ring-4 focus:ring-yellow-400">
                                        <Image
                                        src="/images/k.svg"
                                        alt="logo of the letter k"
                                        height={24}
                                        width={24} />
                                    </a>
                                </Link>
                            </div>
                        <span className="hidden sm:inline-block text-xs text-pri font-bold">
                            Kieran Roberts
                        </span>
                        <a href="https://github.com/kieran6roberts"  
                        className="inline-block text-md text-dark ml-3 transform transition duration-150 ease hover:scale-125 focus:outline-none focus:ring-4 focus:ring-yellow-400">
                            <AiFillGithub />
                        </a>
                        <a href="https://www.linkedin.com/in/kieran-roberts-00517b178/"  
                        className="inline-block text-md text-dark ml-3 transform transition duration-150 ease hover:scale-125 focus:outline-none focus:ring-4 focus:ring-yellow-400">
                            <FaLinkedinIn />
                        </a>
                        <ul className="flex ml-auto font-bold capitalize">
                            <li className="flex-auto bg-gradient-to-r from-purple-400 to-purple-600 p-0.5">
                                <Link href="/#projects" passHref>
                                    <a className="group bg-pri block relative text-xxs text-light text-bold py-2 px-4 md:py-3 md:px-6 overflow-hidden cursor-pointer transition duration-150 ease-in hover:bg-light hover:text-pri focus:outline-none focus:ring-4 focus:ring-yellow-400">
                                        projects
                                        <RiComputerLine className="absolute top-0 left-0 text-lg text-gray-200 text-opacity-30 transition duration-150 ease-in group-hover:text-purple-400 group-hover:text-opacity-30" />
                                    </a>
                                </Link>
                            </li>
                            <li className="flex-auto bg-gradient-to-r from-purple-400 to-purple-600 p-0.5 ml-2 md:ml-4">
                                <Link href="/contact" passHref>
                                    <a className="group bg-pri block relative text-xxs text-light py-2 px-4 md:py-3 md:px-6 overflow-hidden cursor-pointer transition duration-150 ease-in hover:bg-light hover:text-pri focus:outline-none focus:ring-4 focus:ring-yellow-400">
                                        contact me
                                        <AiOutlinePhone className="absolute top-0 left-0 text-lg text-gray-200 text-opacity-30 transition duration-150 ease-in  group-hover:text-purple-400 group-hover:text-opacity-30" />
                                    </a>
                                </Link>
                            </li>
                        </ul>
                    </div>
                </nav>
            </section>
            <main className="mx-3 px-4 md:px-16 border-l-2 border-r-2 border-gray-100 overflow-hidden">
            {children}
            </main>
            <footer className="text-center bg-dark text-light p-4">
                <ul className="flex flex-col items-center">
                    <li className="mb-4">
                        <a href="https://github.com/kieran6roberts"  
                        className="text-md focus:outline-none focus:ring-4 focus:ring-yellow-400">
                            <AiFillGithub className="text-md text-light transform transition duration-150 ease hover:scale-125"/>
                        </a>
                    </li>
                    <li className="mb-4">
                        <a href="https://www.linkedin.com/in/kieran-roberts-00517b178/"  
                        className="text-md focus:outline-none focus:ring-4 focus:ring-yellow-400">
                            <FaLinkedinIn className="text-md text-light transform transition duration-150 ease hover:scale-125"/>
                        </a>
                    </li>
                </ul>
                <p className="uppercase text-sm text-acc">
                    created and designed by Kieran Roberts.
                </p>
                <span className="text-xxs text-gray-400">
                    kieran6roberts@gmail.com
                </span>
            </footer>
        </div>
    )
};