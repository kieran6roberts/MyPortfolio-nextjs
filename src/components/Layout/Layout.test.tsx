import * as React from "react";
import { render, RenderResult } from "@testing-library/react";
import TestRenderer from "react-test-renderer";

import Layout from "@/components/Layout/Layout";

let documentBody: RenderResult;


describe("<Layout />", () => {
    test("renders", () => {
        documentBody = render(<Layout><h1>mock layout header</h1></Layout>);

        expect(documentBody.getByRole("navigation")).toBeInTheDocument();
        expect(documentBody.getByRole("main")).toBeInTheDocument();
        expect(documentBody.getByText(/created and designed by kieran roberts/i)).toBeInTheDocument();

        expect(documentBody.getByRole("link", { name: /Home Kieran Roberts/ })).toBeInTheDocument();
        expect(documentBody.getByRole("link", { name: /projects/ })).toBeInTheDocument();
        expect(documentBody.getByRole("link", { name: /contact me/ })).toBeInTheDocument();
        expect(documentBody.getByRole("heading", { name: /mock layout header/ })).toBeInTheDocument();
    });

    test("render snpashot", () => {
        const tree = TestRenderer
        .create(<Layout><h1>mock layout header</h1></Layout>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});