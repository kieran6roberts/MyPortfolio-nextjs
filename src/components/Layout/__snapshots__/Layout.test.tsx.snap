// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> render snpashot 1`] = `
<div
  className="flex flex-col justify-between min-h-screen font-mono"
>
  <div
    className="w-full h-1 bg-gradient-to-r from-purple-400 to-blue-800"
  />
  <nav
    className="relative z-50 mb-6 shadow-md xl:mb-12 2xl:mb-32"
  >
    <div
      className="flex flex-col items-center px-6 py-2 md:py-6 md:flex-row md:px-16"
    >
      <div
        className="flex p-3 mr-4 cursor-pointer"
      >
        <a
          aria-label="Home Kieran Roberts"
          className="flex align-center focus:outline-none focus:ring-4 focus:ring-yellow-400"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          <span
            className="text-sm font-bold text-pri hover:text-sec"
          >
            Kieran Roberts
          </span>
        </a>
      </div>
      <ul
        className="flex justify-center pt-2 font-bold uppercase md:pt-0 md:ml-auto"
      >
        <li
          className="flex-auto p-0.5"
        >
          <a
            className="relative block px-2 py-2 overflow-hidden transition duration-150 ease-in cursor-pointer md:px-4 group text-xxs lg:text-xs text-sec text-bold lg:py-4 lg:px-8 hover:bg-light hover:text-pri focus:outline-none focus:ring-4 focus:ring-yellow-400"
            href="/#projects"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            projects
            <svg
              className="absolute top-0 left-0 text-purple-200 transition duration-150 ease-in text-opacity-40 text-md sm:text-lg group-hover:text-purple-400 group-hover:text-opacity-30"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M4 16h16V5H4v11zm9 2v2h4v2H7v-2h4v-2H2.992A.998.998 0 0 1 2 16.993V4.007C2 3.451 2.455 3 2.992 3h18.016c.548 0 .992.449.992 1.007v12.986c0 .556-.455 1.007-.992 1.007H13z"
                />
              </g>
            </svg>
          </a>
        </li>
        <li
          className="flex-auto p-0.5 ml-2 md:ml-4"
        >
          <a
            className="relative flex items-center px-2 py-2 overflow-hidden transition duration-150 ease-in cursor-pointer md:px-4 group text-xxs lg:text-xs text-sec lg:py-4 lg:px-8 hover:bg-light hover:text-pri focus:outline-none focus:ring-4 focus:ring-yellow-400"
            href="https://blog.kieranroberts.dev"
            rel="noopener"
            target="_blank"
          >
            blog
            <svg
              className="absolute top-0 left-0 text-purple-200 transition duration-150 ease-in text-md sm:text-lg text-opacity-30 group-hover:text-purple-400 group-hover:text-opacity-30"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20,3H4C2.897,3,2,3.897,2,5v14c0,1.103,0.897,2,2,2h16c1.103,0,2-0.897,2-2V5C22,3.897,21.103,3,20,3z M4,19V7h16 l0.002,12H4z"
              />
              <path
                d="M9.293 9.293L5.586 13 9.293 16.707 10.707 15.293 8.414 13 10.707 10.707zM14.707 9.293L13.293 10.707 15.586 13 13.293 15.293 14.707 16.707 18.414 13z"
              />
            </svg>
            <svg
              className="inline-block mb-1 ml-2"
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
              />
              <polyline
                points="15 3 21 3 21 9"
              />
              <line
                x1="10"
                x2="21"
                y1="14"
                y2="3"
              />
            </svg>
          </a>
        </li>
        <li
          className="flex-auto p-0.5 ml-2 md:ml-4"
        >
          <a
            className="relative block px-2 py-2 overflow-hidden transition duration-150 ease-in cursor-pointer md:px-4 group text-xxs lg:text-xs text-sec lg:py-4 lg:px-8 hover:bg-light hover:text-pri focus:outline-none focus:ring-4 focus:ring-yellow-400"
            href="/contact"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            contact me
            <svg
              className="absolute top-0 left-0 text-purple-200 transition duration-150 ease-in text-md sm:text-lg text-opacity-30 group-hover:text-purple-400 group-hover:text-opacity-30"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M877.1 238.7L770.6 132.3c-13-13-30.4-20.3-48.8-20.3s-35.8 7.2-48.8 20.3L558.3 246.8c-13 13-20.3 30.5-20.3 48.9 0 18.5 7.2 35.8 20.3 48.9l89.6 89.7a405.46 405.46 0 0 1-86.4 127.3c-36.7 36.9-79.6 66-127.2 86.6l-89.6-89.7c-13-13-30.4-20.3-48.8-20.3a68.2 68.2 0 0 0-48.8 20.3L132.3 673c-13 13-20.3 30.5-20.3 48.9 0 18.5 7.2 35.8 20.3 48.9l106.4 106.4c22.2 22.2 52.8 34.9 84.2 34.9 6.5 0 12.8-.5 19.2-1.6 132.4-21.8 263.8-92.3 369.9-198.3C818 606 888.4 474.6 910.4 342.1c6.3-37.6-6.3-76.3-33.3-103.4zm-37.6 91.5c-19.5 117.9-82.9 235.5-178.4 331s-213 158.9-330.9 178.4c-14.8 2.5-30-2.5-40.8-13.2L184.9 721.9 295.7 611l119.8 120 .9.9 21.6-8a481.29 481.29 0 0 0 285.7-285.8l8-21.6-120.8-120.7 110.8-110.9 104.5 104.5c10.8 10.8 15.8 26 13.3 40.8z"
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <main
    className="px-4 mx-3 overflow-hidden border-l-2 border-r-2 border-gray-100 md:px-16"
  >
    <h1>
      mock layout header
    </h1>
  </main>
  <footer
    className="px-4 py-8 text-center 2xl:py-16 2xl:px-8 bg-dark text-light"
  >
    <ul
      className="items-center flex justify-center mb-4 2xl:mb-8"
    >
      <li
        className="mx-3 mb-2 2xl:mx-8"
      >
        <a
          aria-label="to my github page"
          className="text-md"
          href="https://github.com/kieran6roberts"
          rel="noopener"
          target="_blank"
        >
          <svg
            className="text-white transition duration-150 transform text-md ease hover:scale-125"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9a127.5 127.5 0 0 1 38.1 91v112.5c.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z"
            />
          </svg>
        </a>
      </li>
      <li
        className="mx-3 mb-2 2xl:mx-8"
      >
        <a
          aria-label="to my linkedin page"
          className="text-md"
          href="https://www.linkedin.com/in/kieran-roberts-00517b178/"
          rel="noopener"
          target="_blank"
        >
          <svg
            className="text-white transition duration-150 transform text-md ease hover:scale-125"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M100.28 448H7.4V148.9h92.88zM53.79 108.1C24.09 108.1 0 83.5 0 53.8a53.79 53.79 0 0 1 107.58 0c0 29.7-24.1 54.3-53.79 54.3zM447.9 448h-92.68V302.4c0-34.7-.7-79.2-48.29-79.2-48.29 0-55.69 37.7-55.69 76.7V448h-92.78V148.9h89.08v40.8h1.3c12.4-23.5 42.69-48.3 87.88-48.3 94 0 111.28 61.9 111.28 142.3V448z"
            />
          </svg>
        </a>
      </li>
      <li
        className="mx-3 mb-2 2xl:mx-8"
      >
        <a
          aria-label="send me an email"
          className="text-md"
          href="mailto:kieran6roberts@gmail.com?subject=Mail%20from%20portfolio"
          rel="noopener"
          target="_blank"
        >
          <svg
            className="text-white transition duration-150 transform text-md ease hover:scale-125"
            fill="currentColor"
            height="1em"
            role="img"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title />
            <path
              d="M24 4.5v15c0 .85-.65 1.5-1.5 1.5H21V7.387l-9 6.463-9-6.463V21H1.5C.649 21 0 20.35 0 19.5v-15c0-.425.162-.8.431-1.068C.7 3.16 1.076 3 1.5 3H2l10 7.25L22 3h.5c.425 0 .8.162 1.069.432.27.268.431.643.431 1.068z"
            />
          </svg>
        </a>
      </li>
      <li
        className="mx-3 mb-2 2xl:mx-8"
      >
        <a
          aria-label="to my Dev community profile"
          className="text-md"
          href="https://dev.to/kieran6roberts"
          rel="noopener"
          target="_blank"
        >
          <svg
            className="text-white transition duration-150 transform text-md ease hover:scale-125"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M120.12 208.29c-3.88-2.9-7.77-4.35-11.65-4.35H91.03v104.47h17.45c3.88 0 7.77-1.45 11.65-4.35 3.88-2.9 5.82-7.25 5.82-13.06v-69.65c-.01-5.8-1.96-10.16-5.83-13.06zM404.1 32H43.9C19.7 32 .06 51.59 0 75.8v360.4C.06 460.41 19.7 480 43.9 480h360.2c24.21 0 43.84-19.59 43.9-43.8V75.8c-.06-24.21-19.7-43.8-43.9-43.8zM154.2 291.19c0 18.81-11.61 47.31-48.36 47.25h-46.4V172.98h47.38c35.44 0 47.36 28.46 47.37 47.28l.01 70.93zm100.68-88.66H201.6v38.42h32.57v29.57H201.6v38.41h53.29v29.57h-62.18c-11.16.29-20.44-8.53-20.72-19.69V193.7c-.27-11.15 8.56-20.41 19.71-20.69h63.19l-.01 29.52zm103.64 115.29c-13.2 30.75-36.85 24.63-47.44 0l-38.53-144.8h32.57l29.71 113.72 29.57-113.72h32.58l-38.46 144.8z"
            />
          </svg>
        </a>
      </li>
      <li
        className="mx-3 mb-2 2xl:mx-8"
      >
        <a
          aria-label="to my twitter profile"
          className="text-md"
          href="https://twitter.com/Kieran6dev"
          rel="noopener"
          target="_blank"
        >
          <svg
            className="text-white transition duration-150 transform text-md ease hover:scale-125"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M928 254.3c-30.6 13.2-63.9 22.7-98.2 26.4a170.1 170.1 0 0 0 75-94 336.64 336.64 0 0 1-108.2 41.2A170.1 170.1 0 0 0 672 174c-94.5 0-170.5 76.6-170.5 170.6 0 13.2 1.6 26.4 4.2 39.1-141.5-7.4-267.7-75-351.6-178.5a169.32 169.32 0 0 0-23.2 86.1c0 59.2 30.1 111.4 76 142.1a172 172 0 0 1-77.1-21.7v2.1c0 82.9 58.6 151.6 136.7 167.4a180.6 180.6 0 0 1-44.9 5.8c-11.1 0-21.6-1.1-32.2-2.6C211 652 273.9 701.1 348.8 702.7c-58.6 45.9-132 72.9-211.7 72.9-14.3 0-27.5-.5-41.2-2.1C171.5 822 261.2 850 357.8 850 671.4 850 843 590.2 843 364.7c0-7.4 0-14.8-.5-22.2 33.2-24.3 62.3-54.4 85.5-88.2z"
            />
          </svg>
        </a>
      </li>
    </ul>
    <p
      className="my-2 text-xs uppercase 2xl:my-6 text-acc"
    >
      created and designed by Kieran Roberts.
    </p>
    <a
      className="block mx-auto mt-4 2xl:my-6 w-max"
      href="https://graphcms.com/?referrer=kieranroberts.dev"
      rel="noopener"
      target="_blank"
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "display": "inline-block",
            "margin": 0,
            "maxWidth": "100%",
            "overflow": "hidden",
            "position": "relative",
          }
        }
      >
        <div
          style={
            Object {
              "boxSizing": "border-box",
              "display": "block",
              "maxWidth": "100%",
            }
          }
        >
          <img
            alt=""
            aria-hidden={true}
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTIwIiBoZWlnaHQ9IjQwIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIvPg=="
            style={
              Object {
                "display": "block",
                "maxWidth": "100%",
              }
            }
          />
        </div>
        <img
          alt="graphcms logo"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "visibility": "hidden",
              "width": 0,
            }
          }
        />
      </div>
    </a>
    <span
      className="text-gray-400 text-xxs"
    >
      kieran6roberts@gmail.com
    </span>
  </footer>
</div>
`;
